File: TwoActivities/app/src/main/java/com/example/twoactivities/MainActivity.java
package com.example.twoactivities;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    private static final String LOG_TAG = MainActivity.class.getSimpleName();
    public static final String EXTRA_MESSAGE = "com.example.twoactivities.extra.MESSAGE";
    private EditText mMessageEditText;
    private TextView mReplyHeadTextView;
    private TextView mReplyTextView;


    public void openSecondActivity(View view) {
        ActivityResultLauncher<Intent> startForResult = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                new ActivityResultCallback<ActivityResult>()  {
                    @Override
                    public void onActivityResult(ActivityResult result) {
                        if(result != null && result.getResultCode() == RESULT_OK) {
                            if(result.getData() != null && result.getData().getStringExtra(SecondActivity.EXTRA_REPLY) != null){
                                mReplyHeadTextView.setVisibility(View.VISIBLE);
                                mReplyTextView.setText(result.getData().getStringExtra(SecondActivity.EXTRA_REPLY));
                                mReplyTextView.setVisibility(View.VISIBLE);
                            }
                        }
                    }
                });

        Log.d(LOG_TAG, "Button clicked!");
        Intent intent = new Intent(this, SecondActivity.class);
        String message = mMessageEditText.getText().toString();
        intent.putExtra(EXTRA_MESSAGE, message);
        startForResult.launch(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mMessageEditText = findViewById(R.id.editText_main);
        mReplyHeadTextView = findViewById(R.id.text_header_reply);
        mReplyTextView = findViewById(R.id.text_message_reply);

    }

}

File: TwoActivitiesIntentExtras2/app/src/main/java/com/example/twoactivitiesintentextras2/MainActivity.kt
package com.example.twoactivitiesintentextras2

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val button1 = findViewById<Button>(R.id.button1)
        val button2 = findViewById<Button>(R.id.button2)
        val button3 = findViewById<Button>(R.id.button3)

        button1.setOnClickListener{
            launchActivitySecond(1)
        }

        button2.setOnClickListener{
            launchActivitySecond(2)
        }

        button3.setOnClickListener{
            launchActivitySecond(3)
        }
    }

    fun launchActivitySecond(i: Int) {
        val intent = Intent(this, ActivitySecond::class.java)
        intent.putExtra("number", i)
        startActivity(intent)
    }
}